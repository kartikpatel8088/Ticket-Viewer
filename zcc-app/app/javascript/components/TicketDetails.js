

import React,{useEffect, useState} from 'react';
import Home from './Home'
import './App.css';

import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Badge from 'react-bootstrap/Badge';
import Popover from 'react-bootstrap/Popover';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Overlay from 'react-bootstrap/Overlay';

import StarRatingComponent from 'react-star-rating-component';
import moment from 'moment';


/// TicketDetails Class based Component - Stateful

class TicketDetails extends React.Component {
	constructor(props) {
	    super(props);
	    this.state = {
	      details: [],
	      loading: true
	    };
	}
	componentDidMount() {
    	// Runs after the first render() lifecycle
    	const url = "http://localhost:3000/api/v1/ticket/"+this.props.id;
	    	fetch(url, {
			  "method": "GET",
			})
			.then(response => response.json())
			.then(response => {
				this.setState({ details: response.ticket });
				this.setState({ loading: false });
			  	
			  
			})
			.catch(err => { console.log(err); 
			})
	};

  	render() {
    	
    	const popover = (
			  <Popover id="popover-basic">
			    <Popover.Header as="h3">Information</Popover.Header>
			    <Popover.Body>
			      Requested By: {this.state.details.requester_id}<br />
			      Submitted By: {this.state.details.submitter_id} <br />
			      Assigned: {this.state.details.assignee_id} <br />
			      Group: {this.state.details.group_id} <br />
			    </Popover.Body>
			  </Popover>
		);

		const Info = () => (
			  <OverlayTrigger trigger="click" placement="right" overlay={popover}>

			    <Button variant="outline-info"  text="light" size="sm" className="float-right info">Info</Button>
			  </OverlayTrigger>
		);

    	return (
		!this.state.loading ? 
		
			<Modal
		      {...this.props}
		      size="lg"
		      aria-labelledby="contained-modal-title-vcenter"
		      centered
		    >
		    
		      <Modal.Header>
		        <Modal.Title id="contained-modal-title-vcenter">
		          {this.state.details.id}.  {this.state.details.subject}
		          {" "}{this.state.details.status == "open" && <Badge bg="danger" text="light">{this.state.details.status}</Badge>}
		          
		        </Modal.Title>
		        <br />
		        <p className="float-right via">via: {this.state.details.via.channel.toUpperCase()}</p>
		          

		      </Modal.Header>
		      <Modal.Body>
		        <h5>Description</h5>
		        <p className="description">
		          {this.state.details.description}
		        </p>
		        <StarRatingComponent 
		          name="rate1" 
		          editing={false}
		        
		          starCount={5}
		          value={this.state.details.satisfaction_rating}
		        />
		        <br />

		        <Badge pill bg="secondary" text="light">{this.state.details.tags[0]} </Badge>
		        <Badge pill bg="secondary" text="light">{this.state.details.tags[1]} </Badge>
		        <Badge pill bg="secondary" text="light">{this.state.details.tags[2]} </Badge>

		        <Info />
		      </Modal.Body>
		      <Modal.Footer>
		      	<p className="mr-auto">Created On: {moment(this.state.details.created_at).format("LLL")}</p>
		        <Button onClick={this.props.onHide}>Close</Button>
		      </Modal.Footer>
		    </Modal>
		
		:
		<>
		</>
	)
  	}
}


/// TicketDetails Functional based Component - Stateless 

// function TicketDetails(props){
// 	const [details,setDetails]=React.useState([]);
// 	const [loading, setLoading] = React.useState(true);


// 	useEffect(()=>{

// 		const url = "http://localhost:3000/api/v1/ticket/"+props.id;
// 	    	fetch(url, {
// 			  "method": "GET",
// 			})
// 			.then(response => response.json())
// 			.then(response => {
// 			  setDetails(response.ticket); 
// 			  setLoading(false);
// 			})
// 			.catch(err => { console.log(err); 
// 			})

// 	}, []);

// 	const popover = (
// 		  <Popover id="popover-basic">
// 		    <Popover.Header as="h3">Information</Popover.Header>
// 		    <Popover.Body>
// 		      Requested By: {details.requester_id}<br />
// 		      Submitted By: {details.submitter_id} <br />
// 		      Assigned: {details.assignee_id} <br />
// 		      Group: {details.group_id} <br />
// 		    </Popover.Body>
// 		  </Popover>
// 	);

// 	const Example = () => (
// 		  <OverlayTrigger trigger="click" placement="right" overlay={popover}>

// 		    <Button variant="outline-info"  text="light" size="sm" className="float-right">Info</Button>
// 		  </OverlayTrigger>
// 	);

	
	
// 	return (
// 		!loading ? 
		
// 			<Modal
// 		      {...props}
// 		      size="lg"
// 		      aria-labelledby="contained-modal-title-vcenter"
// 		      centered
// 		    >
		    
// 		      <Modal.Header>
// 		        <Modal.Title id="contained-modal-title-vcenter">
// 		          {details.id}.  {details.subject}
// 		          {" "}{details.status == "open" && <Badge bg="danger" text="light">{details.status}</Badge>}
		          
// 		        </Modal.Title>
// 		        <br />
// 		        <p className="float-right via">via: {details.via.channel.toUpperCase()}</p>
		          

// 		      </Modal.Header>
// 		      <Modal.Body>
// 		        <h5>Description</h5>
// 		        <p>
// 		          {details.description}
// 		        </p>
// 		        <StarRatingComponent 
// 		          name="rate1" 
// 		          editing={false}
		        
// 		          starCount={5}
// 		          value={details.satisfaction_rating}
// 		        />
// 		        <br />

// 		        <Badge pill bg="secondary" text="light">{details.tags[0]} </Badge>
// 		        <Badge pill bg="secondary" text="light">{details.tags[1]} </Badge>
// 		        <Badge pill bg="secondary" text="light">{details.tags[2]} </Badge>

// 		        <Example />
// 		      </Modal.Body>
// 		      <Modal.Footer>
// 		      	<p className="mr-auto">Created On: {moment(details.created_at).format("LLL")}</p>
// 		        <Button onClick={props.onHide}>Close</Button>
// 		      </Modal.Footer>
// 		    </Modal>
		
// 		:
// 		<>
// 		</>
// 	)
// }

export default TicketDetails