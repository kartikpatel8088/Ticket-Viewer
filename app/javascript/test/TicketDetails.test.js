
import React from "react";
import { shallow, mount, render } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';
import ReactDOM, {unmountComponentAtNode}from 'react-dom';

import StarRatingComponent from 'react-star-rating-component';
import 'isomorphic-fetch';



import TicketDetails from './../components/TicketDetails';

import Modal from 'react-bootstrap/Modal';

configure({ adapter: new Adapter() });


describe('<TicketDetails />', () => {

	//--------------------------------  Snapshot TESTING: Conditional based component rendering and Integrity Testing ------------------------------------------// 
	
	it('should render initial elements when data is not loaded', () => {
	    const testDetails = {
	      "url": "https://zccstudentskartik.zendesk.com/api/v2/tickets/12.json",
	        "id": 12,
	        "external_id": null,
	        "via": {
	            "channel": "api",
	            "source": {
	                "from": {},
	                "to": {},
	                "rel": null
	            }
	        },
	        "created_at": "2021-11-20T06:22:39Z",
	        "updated_at": "2021-11-20T06:22:39Z",
	        "type": null,
	        "subject": "tempor aliquip sint dolore incididunt",
	        "raw_subject": "tempor aliquip sint dolore incididunt",
	        "description": "Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.\n\nIrure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.",
	        "priority": null,
	        "status": "open",
	        "recipient": null,
	        "requester_id": 1523617346642,
	        "submitter_id": 1523617346642,
	        "assignee_id": 1523617346642,
	        "organization_id": 1900146445865,
	        "group_id": 1900001551265,
	        "collaborator_ids": [],
	        "follower_ids": [],
	        "email_cc_ids": [],
	        "forum_topic_id": null,
	        "problem_id": null,
	        "has_incidents": false,
	        "is_public": true,
	        "due_at": null,
	        "tags": [
	            "ad",
	            "minim",
	            "non"
	        ],
	        "custom_fields": [],
	        "satisfaction_rating": null,
	        "sharing_agreement_ids": [],
	        "fields": [],
	        "followup_ids": [],
	        "ticket_form_id": 1900000972265,
	        "brand_id": 1900000502145,
	        "allow_channelback": false,
	        "allow_attachments": true
	    };

	    const props = {id: 12, show: true}
	    const wrapper = shallow(<TicketDetails {...props} />);
	    
	    wrapper.setState({loading:true, details: testDetails});

	    expect(wrapper.getElements()).toMatchSnapshot();  
  	});
    
	it('should render updated elements when data is loaded', async () => {
	   
	    const props = {id: 12, show: true}

	    const wrapper = await shallow(<TicketDetails {...props} />);

	    const testDetails = {
	        "url": "https://zccstudentskartik.zendesk.com/api/v2/tickets/12.json",
	          "id": 12,
	          "external_id": null,
	          "via": {
	              "channel": "api",
	              "source": {
	                  "from": {},
	                  "to": {},
	                  "rel": null
	              }
	          },
	          "created_at": "2021-11-20T06:22:39Z",
	          "updated_at": "2021-11-20T06:22:39Z",
	          "type": null,
	          "subject": "tempor aliquip sint dolore incididunt",
	          "raw_subject": "tempor aliquip sint dolore incididunt",
	          "description": "Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.\n\nIrure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.",
	          "priority": null,
	          "status": "open",
	          "recipient": null,
	          "requester_id": 1523617346642,
	          "submitter_id": 1523617346642,
	          "assignee_id": 1523617346642,
	          "organization_id": 1900146445865,
	          "group_id": 1900001551265,
	          "collaborator_ids": [],
	          "follower_ids": [],
	          "email_cc_ids": [],
	          "forum_topic_id": null,
	          "problem_id": null,
	          "has_incidents": false,
	          "is_public": true,
	          "due_at": null,
	          "tags": [
	              "ad",
	              "minim",
	              "non"
	          ],
	          "custom_fields": [],
	          "satisfaction_rating": null,
	          "sharing_agreement_ids": [],
	          "fields": [],
	          "followup_ids": [],
	          "ticket_form_id": 1900000972265,
	          "brand_id": 1900000502145,
	          "allow_channelback": false,
	          "allow_attachments": true
	    };

	    wrapper.setState({loading: false, details: testDetails});

	    expect(wrapper.getElements()).toMatchSnapshot();  
	});


  	//-------------------------------- Unit TESTING: Key Feature Testing ------------------------------------------// 

  	it('should check Title contains ticket id and ticket subject', async () => {
	   
	    const props = {id: 12, show: true}

	    const wrapper = await shallow(<TicketDetails {...props} />);

	    const testDetails = {
	        "url": "https://zccstudentskartik.zendesk.com/api/v2/tickets/12.json",
	          "id": 12,
	          "external_id": null,
	          "via": {
	              "channel": "api",
	              "source": {
	                  "from": {},
	                  "to": {},
	                  "rel": null
	              }
	          },
	          "created_at": "2021-11-20T06:22:39Z",
	          "updated_at": "2021-11-20T06:22:39Z",
	          "type": null,
	          "subject": "tempor aliquip sint dolore incididunt",
	          "raw_subject": "tempor aliquip sint dolore incididunt",
	          "description": "Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.\n\nIrure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.",
	          "priority": null,
	          "status": "open",
	          "recipient": null,
	          "requester_id": 1523617346642,
	          "submitter_id": 1523617346642,
	          "assignee_id": 1523617346642,
	          "organization_id": 1900146445865,
	          "group_id": 1900001551265,
	          "collaborator_ids": [],
	          "follower_ids": [],
	          "email_cc_ids": [],
	          "forum_topic_id": null,
	          "problem_id": null,
	          "has_incidents": false,
	          "is_public": true,
	          "due_at": null,
	          "tags": [
	              "ad",
	              "minim",
	              "non"
	          ],
	          "custom_fields": [],
	          "satisfaction_rating": null,
	          "sharing_agreement_ids": [],
	          "fields": [],
	          "followup_ids": [],
	          "ticket_form_id": 1900000972265,
	          "brand_id": 1900000502145,
	          "allow_channelback": false,
	          "allow_attachments": true
	    };

	    wrapper.setState({loading: false, details: testDetails});

	    // console.log(wrapper.find('#contained-modal-title-vcenter'))
	    expect(wrapper.find('ModalTitle').text()).toMatch('12.  tempor aliquip sint dolore incididunt open');  
	});

	it('should check Body contains ticket description', async () => {
	   
	    const props = {id: 12, show: true}

	    const wrapper = await shallow(<TicketDetails {...props} />);

	    const testDetails = {
	        "url": "https://zccstudentskartik.zendesk.com/api/v2/tickets/12.json",
	          "id": 12,
	          "external_id": null,
	          "via": {
	              "channel": "api",
	              "source": {
	                  "from": {},
	                  "to": {},
	                  "rel": null
	              }
	          },
	          "created_at": "2021-11-20T06:22:39Z",
	          "updated_at": "2021-11-20T06:22:39Z",
	          "type": null,
	          "subject": "tempor aliquip sint dolore incididunt",
	          "raw_subject": "tempor aliquip sint dolore incididunt",
	          "description": "Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.\n\nIrure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.",
	          "priority": null,
	          "status": "open",
	          "recipient": null,
	          "requester_id": 1523617346642,
	          "submitter_id": 1523617346642,
	          "assignee_id": 1523617346642,
	          "organization_id": 1900146445865,
	          "group_id": 1900001551265,
	          "collaborator_ids": [],
	          "follower_ids": [],
	          "email_cc_ids": [],
	          "forum_topic_id": null,
	          "problem_id": null,
	          "has_incidents": false,
	          "is_public": true,
	          "due_at": null,
	          "tags": [
	              "ad",
	              "minim",
	              "non"
	          ],
	          "custom_fields": [],
	          "satisfaction_rating": null,
	          "sharing_agreement_ids": [],
	          "fields": [],
	          "followup_ids": [],
	          "ticket_form_id": 1900000972265,
	          "brand_id": 1900000502145,
	          "allow_channelback": false,
	          "allow_attachments": true
	    };

	    wrapper.setState({loading: false, details: testDetails});

	    
	    expect(wrapper.find('.description').text()).toMatch(/Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip/);  
	});


  	it('should check info element exist', async () => {
	   
	    const props = {id: 12, show: true}

	    const wrapper = await shallow(<TicketDetails {...props} />);

	    const testDetails = {
	        "url": "https://zccstudentskartik.zendesk.com/api/v2/tickets/12.json",
	          "id": 12,
	          "external_id": null,
	          "via": {
	              "channel": "api",
	              "source": {
	                  "from": {},
	                  "to": {},
	                  "rel": null
	              }
	          },
	          "created_at": "2021-11-20T06:22:39Z",
	          "updated_at": "2021-11-20T06:22:39Z",
	          "type": null,
	          "subject": "tempor aliquip sint dolore incididunt",
	          "raw_subject": "tempor aliquip sint dolore incididunt",
	          "description": "Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.\n\nIrure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.",
	          "priority": null,
	          "status": "open",
	          "recipient": null,
	          "requester_id": 1523617346642,
	          "submitter_id": 1523617346642,
	          "assignee_id": 1523617346642,
	          "organization_id": 1900146445865,
	          "group_id": 1900001551265,
	          "collaborator_ids": [],
	          "follower_ids": [],
	          "email_cc_ids": [],
	          "forum_topic_id": null,
	          "problem_id": null,
	          "has_incidents": false,
	          "is_public": true,
	          "due_at": null,
	          "tags": [
	              "ad",
	              "minim",
	              "non"
	          ],
	          "custom_fields": [],
	          "satisfaction_rating": null,
	          "sharing_agreement_ids": [],
	          "fields": [],
	          "followup_ids": [],
	          "ticket_form_id": 1900000972265,
	          "brand_id": 1900000502145,
	          "allow_channelback": false,
	          "allow_attachments": true
	    };

	    wrapper.setState({loading: false, details: testDetails});

	    expect(wrapper.find('Info').length).toEqual(1) ;  
	});

	




});


