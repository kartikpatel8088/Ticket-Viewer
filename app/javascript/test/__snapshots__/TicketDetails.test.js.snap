// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketDetails /> should render initial elements when data is not loaded 1`] = `
Array [
  <React.Fragment />,
]
`;

exports[`<TicketDetails /> should render updated elements when data is loaded 1`] = `
Array [
  <Modal
    animation={true}
    aria-labelledby="contained-modal-title-vcenter"
    autoFocus={true}
    backdrop={true}
    centered={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    enforceFocus={true}
    id={12}
    keyboard={true}
    restoreFocus={true}
    show={true}
    size="lg"
  >
    <ModalHeader
      closeButton={false}
      closeLabel="Close"
    >
      <ModalTitle
        id="contained-modal-title-vcenter"
      >
        12
        .  
        tempor aliquip sint dolore incididunt
         
        <Badge
          bg="danger"
          pill={false}
          text="light"
        >
          open
        </Badge>
      </ModalTitle>
      <br />
      <p
        className="float-right via"
      >
        via: 
        API
      </p>
    </ModalHeader>
    <ModalBody>
      <h5>
        Description
      </h5>
      <p
        className="description"
      >
        Amet sint ea minim excepteur amet. Eiusmod duis ad ea ad aliquip occaecat ea voluptate sunt cillum duis aute. Esse in duis tempor et. Est duis nulla irure ullamco amet sit esse Lorem eu aliqua ullamco sunt. In sint dolore id eu ea.

Irure proident nulla deserunt officia laboris culpa. Qui est adipisicing elit tempor cupidatat minim laborum ea est nostrud nulla mollit id dolore. Aute amet amet deserunt incididunt magna ipsum exercitation nostrud incididunt ea qui anim. Adipisicing irure commodo consectetur Lorem ad excepteur nulla eiusmod aliquip irure occaecat. Id sint fugiat proident Lorem est minim ea sint irure. Veniam sint ex minim duis est.
      </p>
      <StarRatingComponent
        editing={false}
        emptyStarColor="#333"
        name="rate1"
        starColor="#ffb400"
        starCount={5}
        value={null}
      />
      <br />
      <Badge
        bg="secondary"
        pill={true}
        text="light"
      >
        ad
         
      </Badge>
      <Badge
        bg="secondary"
        pill={true}
        text="light"
      >
        minim
         
      </Badge>
      <Badge
        bg="secondary"
        pill={true}
        text="light"
      >
        non
         
      </Badge>
      <Info />
    </ModalBody>
    <ModalFooter>
      <p
        className="mr-auto"
      >
        Created On: 
        November 19, 2021 10:22 PM
      </p>
      <Button
        active={false}
        disabled={false}
        variant="primary"
      >
        Close
      </Button>
    </ModalFooter>
  </Modal>,
]
`;
